File: Dart/BSTremove.dart
import "dart:io";

class Node {
  int key;
  Node left, right;

  Node(int item) {
    key = item;
    left = right = null;
  }
}

class BSTremove {
  // Root of BST
  Node root;

  // Constructor
  BSTremove() {
    root = null;
  }

  void insert(int key) {
    root = insertRec(root, key);
  }

  Node insertRec(Node root, int key) {
    // if tree is empty return new node
    if (root == null) {
      root = new Node(key);
      return root;
    }

    // Otherwise traverse down  the tree
    if (key < root.key)
      root.left = insertRec(root.left, key);
    else if (key > root.key) root.right = insertRec(root.right, key);

    return root;
  }

  void deleteKey(int key) {
    root = deleteRec(root, key);
  }

  Node deleteRec(Node root, int key) {
    // base case: tree is empty
    if (root == null) return root;

    // Otherwise traverse down the tree
    if (key < root.key)
      root.left = deleteRec(root.left, key);
    else if (key > root.key)
      root.right = deleteRec(root.right, key);

    // if key is same as root's key, then This is the node
    // to be deleted
    else {
      // node with only one child or no child
      if (root.left == null)
        return root.right;
      else if (root.right == null) return root.left;

      // node with two children: Get the inorder successor (smallest
      // in the right subtree)
      root.key = minValue(root.right);

      // Delete the inorder successor
      root.right = deleteRec(root.right, root.key);
    }

    return root;
  }

  int minValue(Node root) {
    int minv = root.key;
    while (root.left != null) {
      minv = root.left.key;
      root = root.left;
    }
    return minv;
  }
}

void main() {
  var tree = new BSTremove();
  var N = 1000;
  File myFile = new File("array.txt");
  String myFileContent = myFile.readAsStringSync();
  List<String> mylist = myFileContent.split(' ');
  var arr = [];
  for (var i = 0; i < N; i++) arr.add(int.parse(mylist[i]));

  for (var i = 0; i < N; i++) tree.insert(arr[i]);

  Stopwatch stopwatch = new Stopwatch()..start();

  for (var i = 0; i < N; i++) tree.deleteKey(arr[i]);

  print(stopwatch.elapsedMicroseconds); // executing time in microseconds
}

File: c/BSTremove.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <time.h>

#define N 10000

typedef struct node {
    int key;
    struct node *left;
    struct node *right;    
}Node;

Node* newNode (int key);
Node* insert(int newKey, Node* tree);
Node* findNode(int key, Node* tree);
Node* rightmostNode(Node* node);
Node* removeKey(int key, Node* tree);

int main(){
    int arr[N];
    int len = N;
    int num;

    FILE* input = fopen("array.txt", "r");
    for (int i = 0; i < N; i++){
        int s = fscanf(input, "%d", &num);
        arr[i] = num;
    }


    Node *root = NULL;


    for (int i = 0; i < N; i ++){
        root = insert(arr[i], root);
    }
    
    clock_t start, end;
    double insert_time;
    start = clock();
    for (int i = 0; i < N; i++){
        removeKey(arr[i], root);
    }
    end = clock();
    insert_time = ((double) (end - start)) / CLOCKS_PER_SEC;
    printf("%f\n", insert_time * 1000);

    return 0;
}


Node* insert(int newKey, Node* tree){
    if (findNode(newKey, tree) != NULL){
        return tree;
    }

    if (tree == NULL){
        return newNode(newKey);
    }
    if (tree->key > newKey){
        tree->left =  insert(newKey, tree->left);
    }
    else{
        tree->right = insert(newKey, tree->right);
    }
    return tree;
}

Node* findNode(int key, Node* tree){
    Node* current = tree;
    while (current != NULL){
        if (current->key == key){
            return current;
        }
        else if (current->key > key){
            current = current->left;
        }
        else{
            current = current->right;
        }
    }
    return NULL;
}

Node* newNode (int key){
    Node* temp = (Node*) malloc(sizeof(struct node));
    temp->key = key;
    temp->left = NULL;
    temp->right = NULL;
    return temp;
}

Node* rightmostNode(Node* node){
    Node* current = node;
    while (current && current->right != NULL)
        current = current->right;

    return current;
}

Node* removeKey(int key, Node* tree){
    if (tree == NULL){
//        printf("There is no node with key %d\n", key);
        return tree;
    }
    if (tree->key > key){
        tree->left = removeKey(key, tree->left);
    }
    else if (tree->key < key){
        tree->right = removeKey(key, tree->right);
    }
    else{ // node to be deleted is current node

        if (tree->left == NULL){ // node has only right child
            Node *temp = tree->right;
            free(tree);
            return temp;
        }
        else if (tree->right == NULL){  // node has only left child
            Node *temp = tree->left;
            free(tree);
            return temp;
        }
        else{ // node has two children
            // find inorder predeessor, copy the content to the node to be deleted
            // removeKey inorder predecessor
            Node *temp = rightmostNode(tree->left);
            tree->key = temp->key;
            tree->left = removeKey(temp->key, tree->left);
        }
    }
    return tree;
}
